Google Cloud Run Integration

UmungoBungo opened the pull request.

UmungoBungo commented on packages/gcp/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+FROM node:alpine
+
+# ADD ./chromium-x64 /opt/bin/chromium
```
Is it possible to pull Chromium from [lambda-binaries](https://github.com/remotion-dev/lambda-binaries/tree/main/chromium-x64)? At the moment I'm including the whole folder within the repo (on my machine at least)

JonnyBurger replied:
We could theoretically download it from Git, or we also have a S3 bucket which acts as a CDN already for our FFmpeg binaries.

For now it's fine to keep it directly locally

UmungoBungo commented on packages/gcp/src/index.ts:
```diff
@@ -0,0 +1,20 @@
+	// GCP isn't working, perhaps because because index.html references script files with a relative path, and it needs to be referencing these files with a full URL?
+	// At least the AWS url works
+	const comps = await getCompositions(AWSUrl);
+	console.log('comps', comps);
```
when running this function locally, the compositions are returned

UmungoBungo commented on packages/gcp/src/index.ts:
```diff
@@ -0,0 +1,20 @@
+	console.log('comps', comps);
+
+	const browserInstance = await getBrowserInstance(true, {});
+	console.log('browserInstance', browserInstance);
```
when running this locally, it fails on line 47 as it doesn't have Chromium

JonnyBurger replied:
Need to get it from the Lambda binaries and check it in, also the path is hardcoded in Lambda to `/opt/bin/chromium` I think

JonnyBurger replied:
So it should work if the browser is being put in the right path, or the path is adjusted for GCP

UmungoBungo commented:
What I'm currently attempting to solve is to include Chromium within the docker image, and reference that from within `getBrowserInstance()`. I'm not super familiar with containers though, so any help you can give me on the Dockerfile would be great :)

When I run this from Docker, it downloads ffmpeg, ffprobe and Chromium. So it seems that the`PUPPETEER_SKIP_CHROMIUM_DOWNLOAD` flag in the Dockerfile isn't working for me... and I assume ffmpeg and ffprobe are in a similar situation, but I can't actually see where in the code it deciding to download them?

<img width="673" alt="image" src="https://user-images.githubusercontent.com/22192773/214279078-8d150e04-a8bc-4e6e-85be-073db18bffc0.png">

JonnyBurger commented:
We inlined the Puppeteer implementation, so this env variable indeed does not work.
Chrome gets downloaded if is it not found in https://github.com/remotion-dev/remotion/blob/02614d5e69dded70e827289178b10adb98102395/packages/renderer/src/get-local-browser-executable.ts#L23 or the path is not explicitly specified. Guess it looks in `/opt/bin/chromium` and does not find it

UmungoBungo commented:
without @types/react package:
![image](https://user-images.githubusercontent.com/22192773/214759309-26853cfb-8f3b-4e22-ac9f-52c7f75c326e.png)


without webpack package:
![image](https://user-images.githubusercontent.com/22192773/214759269-971efe97-6ed0-4e9f-add9-22da328da3ad.png)

Adding these two packages allows me to move ahead... any ideas?

UmungoBungo commented:
I've realised the issue - I'm not using PNPM in the dockerfile. Trying to fix this atm, advice from pnpm is to make use of the pnpm-lock.yaml file. However, this is at the root of the repo, and from what I've read, Docker is not able to access files from outside of its context.

I spent a little bit of time looking at docker-compose, however the use-case for this is to having multiple containers built at the same time. I also looked at docker context, but the docs state it is there to 'manage multiple Swarm clusters, multiple Kubernetes clusters, and multiple individual Docker nodes.'

Just to keep the ball rolling, i'm going to experiment with defining the dockerfile at the root of the repo, and referencing files from within packages/gcp

UmungoBungo commented:
also I'm having issues adding the chromium binary to the repo, as it is complaining about something to do with LFS. I haven't looked into this in detail yet - did you have to work around this for the binaries repo?

UmungoBungo commented:
ok, had some success with moving the dockerfile to the root in order to access the pnpm-lock.yaml. I have been able to use pnpm, however there are issues that occur due to turborepo (I think), [here is the terminal output](https://app.warp.dev/block/Bb1GI7WNvDqS4GN51rrZqe)

There is a [fairly long blogpost](https://app.warp.dev/block/Bb1GI7WNvDqS4GN51rrZqe) I need to go through, where the author manages to not re-download packages on subsequent builds of his Docker containers.

Any ideas or comments would be great!

JonnyBurger commented:
Let's see if we can take some shortcuts to move forward..

- FFmpeg and Puppeteer can also be installed from apt inside a Dockerfile, maybe this can act as a starting point: https://github.com/remotion-dev/template-helloworld/blob/b011dd85261803cdec7e2579ca34d67c02f2dc75/Dockerfile
While less flexibility than providing our own binaries, and not having codecs this should be a quick solution for now.

- Using the files from the monorepo directly is anyway a bit prone to problems, because what's inside the monorepo is not the same as what's being published to npm, and that's what the user is eventually going to use.

What we do for Lambda is make a .zip bundle: https://github.com/remotion-dev/remotion/blob/main/packages/lambda/src/admin/bundle-lambda.ts

This also results in less code being used inside Lambda, since the code is stripped of node_modules.

UmungoBungo commented:
I tried a few more attempts with getting the pnpm build step to work within the docker image, but didn't make much progress there. Instead, I moved gcp out to its own folder and I am just using npm there. Not intending for this to be final of course :)

I've added chromium and ffmpeg in the docker build steps, and I can tell this is being recognised in the code because if I don't add them, the first step when running the docker container is to download them both. It still downloads ffprobe, though I couldn't see a binary for that..?

The error that I haven't been able to overcome is launching the browser itself. I am just running this container locally, I still haven't moved into GCP yet :(
Here is the error message
![image](https://user-images.githubusercontent.com/22192773/215047954-d15c7eef-a13a-4687-82de-3abcc7f1be7a.png)

If I don't add chromium in the docker build step, the message changes, so at least we know that Chrome is actually there, it is just having issues launching...
![image](https://user-images.githubusercontent.com/22192773/215048140-9bad1662-2e4c-4071-8b31-f9687ad9d58f.png)



I have tried changing the node version from alpine to 19.5, in case something is missing, but that hasn't helped at all.

JonnyBurger commented on gcpStandalone/src/index.ts:
```diff
@@ -0,0 +1,34 @@
+ff.http('TypescriptFunction', async (req: ff.Request, res: ff.Response) => {
+	const AWSUrl =
+		'https://remotionlambda-11kow3vq6f.s3.us-east-1.amazonaws.com/sites/xmycbufjs3/index.html';
+	const GCPUrl =
```
When you bundle the site, you need to pass `publicPath` as `/remotionlambda-test/sites/xmycbufjs3/`, otherwise the bundle.js will be loaded from the root of the domain

alexfernandez803 commented:
@UmungoBungo  do you need help in this one? I was able to build the docker image with chrome.

UmungoBungo commented:
thanks @alexfernandez803 I'm all good for now, will reach out if I need a hand 🤝

UmungoBungo commented:
Some success 🎉 

Bundling the code into a single `index.js` in a new folder, `container/`.
The reason for the new folder is that I am using the [`functions-framework`](https://github.com/GoogleCloudPlatform/functions-framework-nodejs) package to expose the code in a local environment or to run it in a cloud function or cloud run environment.

Chromium and FFMPEG are being installed via the dockerfile, and I can see that Chromium is available at `/usr/bin/chromium-browser`. I've updated the path in `get-chromium-executable-path` as well. Here is the output from Docker, which I still need to read through in detail to try and work out what the next steps are.

<details>
  <summary>Docker Output</summary>
  
  ```
  2023-01-31 11:39:55 
2023-01-31 11:39:55 > gcp-container@0.0.1 start
2023-01-31 11:39:55 > functions-framework --target=functionStart
2023-01-31 11:39:55 
2023-01-31 11:39:56 Serving function...
2023-01-31 11:39:56 Function: functionStart
2023-01-31 11:39:56 Signature type: http
2023-01-31 11:39:56 URL: http://localhost:8080/
2023-01-31 11:40:03 [0131/034003.082683:WARNING:dns_config_service_linux.cc(429)] Failed to read DnsConfig.
2023-01-31 11:40:03 [0131/034003.082729:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /var/run/dbus/system_bus_socket: No such file or directory
2023-01-31 11:40:03 [0131/034003.082824:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /var/run/dbus/system_bus_socket: No such file or directory
2023-01-31 11:40:03 [0131/034003.092782:WARNING:audio_manager_linux.cc(60)] Falling back to ALSA for audio output. PulseAudio is not available or could not be initialized.
2023-01-31 11:40:03 
2023-01-31 11:40:03 DevTools listening on ws://127.0.0.1:35701/devtools/browser/068042c1-1c27-4db3-9ad9-c0e8014d3a8a
2023-01-31 11:40:03 [0131/034003.106501:WARNING:bluez_dbus_manager.cc(247)] Floss manager not present, cannot set Floss enable/disable.
2023-01-31 11:40:03 [0131/034003.110202:ERROR:angle_platform_impl.cc(43)] RendererVk.cpp:138 (VerifyExtensionsPresent): Extension not supported: VK_KHR_surface
2023-01-31 11:40:03 ERR: RendererVk.cpp:138 (VerifyExtensionsPresent): Extension not supported: VK_KHR_surface
2023-01-31 11:40:03 [0131/034003.110234:ERROR:angle_platform_impl.cc(43)] RendererVk.cpp:138 (VerifyExtensionsPresent): Extension not supported: VK_KHR_xcb_surface
2023-01-31 11:40:03 ERR: RendererVk.cpp:138 (VerifyExtensionsPresent): Extension not supported: VK_KHR_xcb_surface
2023-01-31 11:40:03 [0131/034003.110269:ERROR:angle_platform_impl.cc(43)] Display.cpp:1021 (initialize): ANGLE Display::initialize error 0: Internal Vulkan error (-7): A requested extension is not supported, in ../../third_party/angle/src/libANGLE/renderer/vulkan/RendererVk.cpp, initialize:1574.
2023-01-31 11:40:03 ERR: Display.cpp:1021 (initialize): ANGLE Display::initialize error 0: Internal Vulkan error (-7): A requested extension is not supported, in ../../third_party/angle/src/libANGLE/renderer/vulkan/RendererVk.cpp, initialize:1574.
2023-01-31 11:40:03 [0131/034003.110293:ERROR:gl_display.cc(508)] EGL Driver message (Critical) eglInitialize: Internal Vulkan error (-7): A requested extension is not supported, in ../../third_party/angle/src/libANGLE/renderer/vulkan/RendererVk.cpp, initialize:1574.
2023-01-31 11:40:03 [0131/034003.110434:ERROR:gl_display.cc(932)] eglInitialize SwANGLE failed with error EGL_NOT_INITIALIZED
2023-01-31 11:40:03 [0131/034003.110455:ERROR:gl_ozone_egl.cc(23)] GLDisplayEGL::Initialize failed.
2023-01-31 11:40:03 [0131/034003.139056:ERROR:gpu_channel_manager.cc(884)] Failed to create GLES3 context, fallback to GLES2.
2023-01-31 11:40:03 [0131/034003.139080:ERROR:gpu_channel_manager.cc(895)] ContextResult::kFatalFailure: Failed to create shared context for virtualization.
2023-01-31 11:40:03 [0131/034003.139089:ERROR:shared_image_stub.cc(439)] SharedImageStub: unable to create context
2023-01-31 11:40:03 [0131/034003.139153:ERROR:gpu_channel.cc(589)] GpuChannel: Failed to create SharedImageStub
2023-01-31 11:40:03 [0131/034003.139252:ERROR:gpu_channel_manager.cc(884)] Failed to create GLES3 context, fallback to GLES2.
2023-01-31 11:40:03 [0131/034003.139265:ERROR:gpu_channel_manager.cc(895)] ContextResult::kFatalFailure: Failed to create shared context for virtualization.
2023-01-31 11:40:03 [0131/034003.139273:ERROR:shared_image_stub.cc(439)] SharedImageStub: unable to create context
2023-01-31 11:40:03 [0131/034003.139286:ERROR:gpu_channel.cc(589)] GpuChannel: Failed to create SharedImageStub
2023-01-31 11:40:03 [0131/034003.151436:ERROR:gpu_channel_manager.cc(884)] Failed to create GLES3 context, fallback to GLES2.
2023-01-31 11:40:03 [0131/034003.151459:ERROR:gpu_channel_manager.cc(895)] ContextResult::kFatalFailure: Failed to create shared context for virtualization.
2023-01-31 11:40:03 [0131/034003.151468:ERROR:shared_image_stub.cc(439)] SharedImageStub: unable to create context
2023-01-31 11:40:03 [0131/034003.151479:ERROR:gpu_channel.cc(589)] GpuChannel: Failed to create SharedImageStub
2023-01-31 11:40:03 browserInstance Browser {
2023-01-31 11:40:03   eventsMap: Map(4) {
2023-01-31 11:40:03     'targetcreated' => [],
2023-01-31 11:40:03     'targetchanged' => [],
2023-01-31 11:40:03     'closed' => [ [Function: onceHandler] ],
2023-01-31 11:40:03     'disconnected' => [ [Function (anonymous)] ]
2023-01-31 11:40:03   },
2023-01-31 11:40:03   emitter: {
2023-01-31 11:40:03     all: Map(4) {
2023-01-31 11:40:03       'targetcreated' => [],
2023-01-31 11:40:03       'targetchanged' => [],
2023-01-31 11:40:03       'closed' => [Array],
2023-01-31 11:40:03       'disconnected' => [Array]
2023-01-31 11:40:03     },
2023-01-31 11:40:03     on: [Function: on],
2023-01-31 11:40:03     off: [Function: off],
2023-01-31 11:40:03     emit: [Function: emit]
2023-01-31 11:40:03   },
2023-01-31 11:40:03   connection: Connection {
2023-01-31 11:40:03     eventsMap: Map(3) {
2023-01-31 11:40:03       'Target.targetCreated' => [Array],
2023-01-31 11:40:03       'Target.targetDestroyed' => [Array],
2023-01-31 11:40:03       'Target.targetInfoChanged' => [Array]
2023-01-31 11:40:03     },
2023-01-31 11:40:03     emitter: {
2023-01-31 11:40:03       all: [Map],
2023-01-31 11:40:03       on: [Function: on],
2023-01-31 11:40:03       off: [Function: off],
2023-01-31 11:40:03       emit: [Function: emit]
2023-01-31 11:40:03     }
2023-01-31 11:40:03   }
2023-01-31 11:40:03 }
2023-01-31 11:40:03 The path exists.

  ```
</details>

UmungoBungo commented:
Deployed to Cloud Run, same messages as I was getting locally. Have updated Readme with instructions for deployment, but will improve these over time.

JonnyBurger commented:
Nice! I think Chrome warnings can be ignored, it will work anyway. It's noisy usually at least...

Is the plan to let the users customize the code inside the container or ship it binary format like Lambda? We have to make this decision soon I guess.

UmungoBungo commented:
Render Still is packaged up and working on Cloud Run. A bit of work to do on architecture, but it is working
`https://render-still-tv4h6muxbq-uc.a.run.app/?composition=StillRender`

![image](https://user-images.githubusercontent.com/22192773/216233066-1b97c9cc-3268-45ed-a636-ffc736d8a5fe.png)

UmungoBungo commented:
Render Media also working. I had to bump the Cloud Run memory up to 1gb:
![image](https://user-images.githubusercontent.com/22192773/216236532-2de31c74-e0be-4733-8340-205f900e54d1.png)

Going to `https://render-media-tv4h6muxbq-uc.a.run.app/?composition=HelloWorld` creates this in the storage bucket

https://user-images.githubusercontent.com/22192773/216236570-5c6cc7c1-f42d-4745-9d3f-8ac5090bccf0.mp4

JonnyBurger commented:
Very nice!! Can you push the changes?

UmungoBungo commented:
reconfigured the code to now only have one container created, and you can pass in a `type` prop to render either a still or media. Updates the readme at the root of this package with details

JonnyBurger commented on packages/gcp/readme.md:
```diff
@@ -0,0 +1,167 @@
+Options:
+
+- --project-id (required):
+  - The project-id is the name of the project in which to deploy the service. The ID can be retrieved by navigating to the [Dashboard](https://console.cloud.google.com/home/dashboard?project=new-remotion-project&supportedpurview=project) screen in Google Cloud Console, and selecting the project from the drop-down menu in the top left corner. The ID is under Project Info in the top left corner.
```
The URL here seems to be hardcoded to new-remotion-project

JonnyBurger replied:
The URL here seems to be hardcoded to `new-remotion-project`

JonnyBurger commented on packages/gcp/src/shared/validate-project-id.ts:
```diff
@@ -0,0 +1,35 @@
+	// TODO: Complete this with the node API instead of gcloud
+	exec(`gcloud projects describe ${projectID}`,
+		(error) => {
+			if (error !== null) {
```
I got to this case but due the following error:

  WARNING: Could not open the configuration file: [/Users/jonathanburger/.config/gcloud/configurations/config_default].
  ERROR: (gcloud.projects.describe) You do not currently have an active account selected.

So yes it is best to kill it and use the Node SDK

UmungoBungo commented on packages/gcp/readme.md:
```diff
@@ -0,0 +1,167 @@
+- --service-name (required):
+  - the name of the Cloud Run service to deploy or update. Service names must be 49 characters or less and must be unique per region and project. A service name cannot be changed later and is publicly visible. The service does not need to exist yet. If it does exist, a new revision will be deployed.
+- --region: the region of the service
+- --remotion-version (required for now): the version of the remotion package to use
```
oh yeh you can forget the remotion-version now, I made this comment when only `3.3.36-alpha` was available in the artifact registry. I'll update the readme

UmungoBungo replied:
if you leave it out, it just grabs the image that matches the latest version number. Cool that it worked!

JonnyBurger commented on packages/cloudrun/install.mjs:
```diff
@@ -0,0 +1,201 @@
+		});
+
+		rl.question(
+			'What version of Remotion do you want to use (format: 1.0.0)? \u001b[34;1m',
```
Let's definitely default it to latest (getLatestRemotionVersion() function could be copied from somewhere else

JonnyBurger commented on packages/cloudrun/install.mjs:
```diff
@@ -0,0 +1,606 @@
+	 * Check for existing Terraform State
+	 ****************************************/
+
+	const dirFiles = execSync('ls -a', {
```
this could also use `fs.readdirSync()`, probably easier to parse since it returns an array

JonnyBurger commented:
@SocketSecurity ignore create-video@3.3.94
@SocketSecurity ignore color-names@1.4.0

JonnyBurger commented:
@SocketSecurity ignore color-namer@1.4.0
